File: core/src/main/java/com/ustadmobile/core/controller/ClazzLogListPresenter.java
package com.ustadmobile.core.controller;

import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.db.dao.ClazzDao;
import com.ustadmobile.core.db.dao.ClazzLogAttendanceRecordDao;
import com.ustadmobile.core.db.dao.ClazzMemberDao;
import com.ustadmobile.core.db.dao.ThresholdResult;
import com.ustadmobile.core.generated.locale.MessageID;
import com.ustadmobile.core.impl.UmAccountManager;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.core.impl.UmCallbackWithDefaultValue;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.core.util.UMCalendarUtil;
import com.ustadmobile.core.view.ClassLogDetailView;
import com.ustadmobile.core.view.ClassLogListView;
import com.ustadmobile.lib.db.entities.ClazzLog;
import com.ustadmobile.lib.db.entities.ClazzLogWithScheduleStartEndTimes;
import com.ustadmobile.lib.db.entities.DailyAttendanceNumbers;
import com.ustadmobile.lib.db.entities.Role;

import java.util.Calendar;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static com.ustadmobile.core.view.ClazzListView.ARG_CLAZZ_UID;
import static com.ustadmobile.core.view.ClassLogListView.CHART_DURATION_LAST_MONTH;
import static com.ustadmobile.core.view.ClassLogListView.CHART_DURATION_LAST_WEEK;
import static com.ustadmobile.core.view.ClassLogListView.CHART_DURATION_LAST_YEAR;
import static com.ustadmobile.lib.db.entities.ClazzLogAttendanceRecord.STATUS_ATTENDED;

/**
 * The Presenter/Controller for ClazzLogListFragment. This is responsible for the logic behind
 * displaying every Clazz Attendance Logs via the database provider from the database/repository.
 * It is also responsible for the logic for displaying attendance charts for that class.
 */
public class ClazzLogListPresenter extends UstadBaseController<ClassLogListView>{

    private long currentClazzUid = 0L;

    private UmProvider<ClazzLogWithScheduleStartEndTimes> clazzLogListProvider;

    UmAppDatabase repository = UmAccountManager.getRepositoryForActiveAccount(context);
    private Long loggedInPersonUid;

    public ClazzLogListPresenter(Object context, Hashtable arguments, ClassLogListView view) {
        super(context, arguments, view);

        //Get clazz uid and set it
        if(arguments.containsKey(ARG_CLAZZ_UID)){
            currentClazzUid = (long) arguments.get(ARG_CLAZZ_UID);
        }

        loggedInPersonUid = UmAccountManager.getActiveAccount(context).getPersonUid();

        //Permissions
        checkPermissions();
    }

    public void checkPermissions(){
        ClazzDao clazzDao = repository.getClazzDao();
        clazzDao.personHasPermission(loggedInPersonUid, currentClazzUid,
                Role.PERMISSION_CLAZZ_LOG_ATTENDANCE_INSERT,
                new UmCallbackWithDefaultValue<>(false, new UmCallback<Boolean>() {
            @Override
            public void onSuccess(Boolean result) {
                view.setFABVisibility(result);
            }

            @Override
            public void onFailure(Throwable exception) {
                exception.printStackTrace();
            }
        }));
    }

    /**
     * In Order:
     *      1. This populates the Attendance/Log Entry list provider and sets it to the View.
     *      2. Generate attendance line chart
     *      3. Generate attendance bar chart
     *
     * This will be called when the implementation's View is ready.
     * (ie: on Android, this is called in the Fragment's onCreateView() )
     *
     * @param savedState    The savedState
     */
    @Override
    public void onCreate(Hashtable savedState){
        super.onCreate(savedState);

        clazzLogListProvider =
                repository.getClazzLogDao().findByClazzUidNotCancelledWithSchedule(currentClazzUid);
        setProviderToView();

    }

    /**
     * Sets the clazz log list / clazz attendance list UMProvider provider set in this Presenter
     * to the View object.
     */
    private void setProviderToView(){
        view.setClazzLogListProvider(clazzLogListProvider);
    }

    /**
     * Method logic to go to the Log Detail activity - To see the attendance entry details and edit
     * them.
     *
     * @param clazzLog  The clazz log entry uid that we want to edit.
     */
    public void goToClazzLogDetailActivity(ClazzLog clazzLog){

        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        Hashtable<String, Object> args = new Hashtable<>();
        args.put(ClassLogDetailView.ARG_CLAZZ_LOG_UID, String.valueOf(clazzLog.getClazzLogUid()));
        impl.go(ClassLogDetailView.VIEW_NAME, args, view.getContext());
    }

    /**
     *Method logic that goes to ClazzLogDetail view (take attendance) for the class we're in.
     */
    public void goToNewClazzLogDetailActivity(){

        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        repository.getClazzLogDao().findMostRecentByClazzUid(currentClazzUid, new UmCallback<ClazzLog>() {
            @Override
            public void onSuccess(ClazzLog result) {
                if(result == null){
                    view.showMessage(MessageID.no_schedule_message);
                }else{
                    Hashtable<String, Object> args = new Hashtable<>();
                    args.put(ClassLogDetailView.ARG_MOST_RECENT_BY_CLAZZ_UID,
                            String.valueOf(currentClazzUid));
                    impl.go(ClassLogDetailView.VIEW_NAME, args, view.getContext());
                }
            }

            @Override
            public void onFailure(Throwable exception) {
                exception.printStackTrace();
            }
        });

    }

    /**
     * Method that takes the duration flag and calculated daily attendance numbers for the current
     * class and updates both the line and bar charts.
     *
     * @param duration The duration flag (CHART_DURATION_LAST_WEEK, CHART_DURATION_LAST_MONTH,
     *                 CHART_DURATION_LAST_YEAR) as per defined in ClassLogListView
     */
    public void getAttendanceDataAndUpdateCharts(int duration){

        //Reset report buttons
        view.resetReportButtons();

        LinkedHashMap<Float, Float> lineDataMap = new LinkedHashMap<>();
        LinkedHashMap<Float, Float> barDataMap = new LinkedHashMap<>();
        long toDate = System.currentTimeMillis();
        long fromDate = toDate;

        switch (duration){
            case CHART_DURATION_LAST_WEEK:
                //7
                fromDate = UMCalendarUtil.getDateInMilliPlusDays(-7);
                for(int i=-7; i<-1; i++) {
                    lineDataMap.put((float) UMCalendarUtil.getDateInMilliPlusDays(i) / 1000, -1f);
                }
                break;
            case CHART_DURATION_LAST_MONTH:
                //31
                fromDate = UMCalendarUtil.getDateInMilliPlusDays(-31);
                for(int i=-31; i<-1; i++){
                    lineDataMap.put((float) UMCalendarUtil.getDateInMilliPlusDays(i) / 1000, -1f);
                }
                break;
            case CHART_DURATION_LAST_YEAR:
                //31
                fromDate = UMCalendarUtil.getDateInMilliPlusDays(-365);
                for(int i=-365; i<-1; i++){
                    lineDataMap.put((float) UMCalendarUtil.getDateInMilliPlusDays(i) / 1000, -1f);
                }
                break;
            default:
                //Do nothing.
                break;
        }

        //Calculate daily attendance numbers from the database for the line chart.
        ClazzLogAttendanceRecordDao attendanceRecordDao =
                repository.getClazzLogAttendanceRecordDao();
        attendanceRecordDao.findDailyAttendanceByClazzUidAndDateAsync( currentClazzUid, fromDate,
                toDate, new UmCallback<List<DailyAttendanceNumbers>>() {
            @Override
            public void onSuccess(List<DailyAttendanceNumbers> result) {

                for(DailyAttendanceNumbers everyDayAttendance: result){
                    long dd =everyDayAttendance.getLogDate();
                    Calendar calendar = Calendar.getInstance();
                    calendar.setTimeInMillis(dd);
                    calendar.set(Calendar.HOUR_OF_DAY, 0);
                    calendar.set(Calendar.MINUTE, 0);
                    calendar.set(Calendar.SECOND, 0);
                    calendar.set(Calendar.MILLISECOND, 0);
                    Long d = calendar.getTimeInMillis();
                    float a = everyDayAttendance.getAttendancePercentage();
                    lineDataMap.put(d.floatValue() / 1000, a);

                }

                //Remove messy date keys
                Iterator<Map.Entry<Float, Float>> ldpi = lineDataMap.entrySet().iterator();
                LinkedHashMap<Float, Float> lineDataMapFixedX = new LinkedHashMap<>();
                float l = 0f;
                while(ldpi.hasNext()){
                    l++;
                    lineDataMapFixedX.put(l, ldpi.next().getValue());
                }

                view.updateAttendanceLineChart(lineDataMapFixedX);

            }

            @Override
            public void onFailure(Throwable exception) {
                exception.printStackTrace();
            }
        });

        //Calculate attendance average numbers for the bar chart.
        ClazzMemberDao clazzMemberDao = repository.getClazzMemberDao();

        clazzMemberDao.findAttendanceSpreadByThresholdForTimePeriodAndClazzAndType(STATUS_ATTENDED,
                currentClazzUid, fromDate, toDate, new UmCallback<ThresholdResult>() {
                    @Override
                    public void onSuccess(ThresholdResult result) {
                        barDataMap.put(3f, result.getHigh()/100);
                        barDataMap.put(2f, result.getMid()/100);
                        barDataMap.put(1f, result.getLow()/100);
                        view.updateAttendanceBarChart(barDataMap);
                    }

                    @Override
                    public void onFailure(Throwable exception) {
                        exception.printStackTrace();
                    }
                });

    }
}

File: core/src/commonMain/kotlin/com/ustadmobile/core/controller/ClazzLogListPresenter.kt
package com.ustadmobile.core.controller


import androidx.paging.DataSource
import com.ustadmobile.core.generated.locale.MessageID
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.core.util.UMCalendarUtil
import com.ustadmobile.core.view.ClassLogDetailView
import com.ustadmobile.core.view.ClassLogListView
import com.ustadmobile.core.view.ClassLogListView.Companion.CHART_DURATION_LAST_MONTH
import com.ustadmobile.core.view.ClassLogListView.Companion.CHART_DURATION_LAST_WEEK
import com.ustadmobile.core.view.ClassLogListView.Companion.CHART_DURATION_LAST_YEAR
import com.ustadmobile.core.view.ClazzListView.Companion.ARG_CLAZZ_UID
import com.ustadmobile.lib.db.entities.ClazzLog
import com.ustadmobile.lib.db.entities.ClazzLogAttendanceRecord.Companion.STATUS_ATTENDED
import com.ustadmobile.lib.db.entities.ClazzLogWithScheduleStartEndTimes
import com.ustadmobile.lib.db.entities.Role
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

/**
 * The Presenter/Controller for ClazzLogListFragment. This is responsible for the logic behind
 * displaying every Clazz Attendance Logs via the database provider from the database/repository.
 * It is also responsible for the logic for displaying attendance charts for that class.
 */
class ClazzLogListPresenter(context: Any, arguments: Map<String, String>?, view: ClassLogListView,
                            val impl : UstadMobileSystemImpl = UstadMobileSystemImpl.instance)
    : UstadBaseController<ClassLogListView>(context, arguments!!, view) {

    private var currentClazzUid = 0L

    private var clazzLogListProvider: DataSource.Factory<Int, ClazzLogWithScheduleStartEndTimes>? =
            null

    internal var repository = UmAccountManager.getRepositoryForActiveAccount(context)
    private val loggedInPersonUid: Long?

    init {

        //Get clazz uid and set it
        if (arguments!!.containsKey(ARG_CLAZZ_UID)) {
            currentClazzUid = arguments!!.get(ARG_CLAZZ_UID)!!.toLong()
        }

        loggedInPersonUid = UmAccountManager.getActiveAccount(context)!!.personUid

        //Permissions
        checkPermissions()
    }

    /**
     * Check permission and update the view accordingly
     */
    fun checkPermissions() {
        val clazzDao = repository.clazzDao
        GlobalScope.launch {
            var result = clazzDao.personHasPermission(loggedInPersonUid!!, currentClazzUid,
                    Role.PERMISSION_CLAZZ_LOG_ATTENDANCE_INSERT)
            if(result == null){
                result = false
            }
            view.setFABVisibility(result)
        }
    }

    /**
     * In Order:
     * 1. This populates the Attendance/Log Entry list provider and sets it to the View.
     * 2. Generate attendance line chart
     * 3. Generate attendance bar chart
     *
     * This will be called when the implementation's View is ready.
     * (ie: on Android, this is called in the Fragment's onCreateView() )
     *
     * @param savedState    The savedState
     */
    override fun onCreate(savedState: Map<String, String?>?) {
        super.onCreate(savedState)
        GlobalScope.launch {
            clazzLogListProvider =
                    repository.clazzLogDao.findByClazzUidNotCancelledWithSchedule(currentClazzUid)
            setProviderToView()
        }

    }

    /**
     * Sets the clazz log list / clazz attendance list UMProvider provider set in this Presenter
     * to the View object.
     */
    private fun setProviderToView() {
        view.setClazzLogListProvider(clazzLogListProvider!!)
    }

    /**
     * Method logic to go to the Log Detail activity - To see the attendance entry details and edit
     * them.
     *
     * @param clazzLog  The clazz log entry uid that we want to edit.
     */
    fun goToClazzLogDetailActivity(clazzLog: ClazzLog) {

        val args = HashMap<String, String>()
        args.put(ClassLogDetailView.ARG_CLAZZ_LOG_UID, clazzLog.clazzLogUid.toString())
        impl.go(ClassLogDetailView.VIEW_NAME, args, view.viewContext)
    }

    /**
     * Method logic that goes to ClazzLogDetail view (take attendance) for the class we're in.
     */
    fun goToNewClazzLogDetailActivity() {


        GlobalScope.launch {
            val result = repository.clazzLogDao.findMostRecentByClazzUid(currentClazzUid)

            if (result == null) {
                //Check if no schedule
                val clazzSchedules =
                        repository.scheduleDao.findAllSchedulesByClazzUidAsync(currentClazzUid)
                if(clazzSchedules.isEmpty()) {
                    view.showMessage(MessageID.no_schedule_message)
                }else{
                    //TODO
                    //Find next scheduled attendance
//                    val today = UMCalendarUtil.getDayOfWeek(UMCalendarUtil.getDateInMilliPlusDays(0)) + 1
//                    for(s in clazzSchedules){
//                        val sday = s.scheduleDay
//                        if(sday != 0){ //Weekly
//
//                            //TODO: Get number of days
//                            val days = 42
//                            val message = impl.getString(MessageID.next_attendance_message, view.viewContext) +
//                                    " " + impl.getString(MessageID.it_will_start_in, view.viewContext) + " "  +
//                                    days + " " + impl.getString(MessageID.days, view.viewContext)
//                            view.showMessage(message)
//
//
//                        }else{
//
//                            val message = impl.getString(MessageID.next_attendance_message, view.viewContext) +
//                                    " " + impl.getString(MessageID.it_will_start_tomorrow, view.viewContext) + ""
//                            view.showMessage(message)
//                        }
//                    }

                    view.showMessage(MessageID.next_attendance_message)

                }
            } else {
                val args = HashMap<String, String>()
                args.put(ClassLogDetailView.ARG_MOST_RECENT_BY_CLAZZ_UID,
                        currentClazzUid.toString())
                impl.go(ClassLogDetailView.VIEW_NAME, args, view.viewContext)
            }
        }

    }

    /**
     * Method that takes the duration flag and calculated daily attendance numbers for the current
     * class and updates both the line and bar charts.
     *
     * @param duration The duration flag (CHART_DURATION_LAST_WEEK, CHART_DURATION_LAST_MONTH,
     * CHART_DURATION_LAST_YEAR) as per defined in ClassLogListView
     */
    fun getAttendanceDataAndUpdateCharts(duration: Int) {

        //Reset report buttons
        view.resetReportButtons()

        val lineDataMap = LinkedHashMap<Float, Float>()
        val barDataMap = LinkedHashMap<Float, Float>()
        val toDate = UMCalendarUtil.getDateInMilliPlusDays(0)
        var fromDate = toDate

        when (duration) {
            CHART_DURATION_LAST_WEEK -> {
                //7
                fromDate = UMCalendarUtil.getDateInMilliPlusDays(-7)
                for (i in -7 until -1) {
                    lineDataMap[UMCalendarUtil.getDateInMilliPlusDays(i).toFloat() / 1000] = -1f
                }
            }
            CHART_DURATION_LAST_MONTH -> {
                //31
                fromDate = UMCalendarUtil.getDateInMilliPlusDays(-31)
                for (i in -31 until -1) {
                    lineDataMap[UMCalendarUtil.getDateInMilliPlusDays(i).toFloat() / 1000] = -1f
                }
            }
            CHART_DURATION_LAST_YEAR -> {
                //31
                fromDate = UMCalendarUtil.getDateInMilliPlusDays(-365)
                for (i in -365 until -1) {
                    lineDataMap[UMCalendarUtil.getDateInMilliPlusDays(i).toFloat() / 1000] = -1f
                }
            }
            else -> {
            }
        }//Do nothing.

        //Calculate daily attendance numbers from the database for the line chart.
        val attendanceRecordDao = repository.clazzLogAttendanceRecordDao
        GlobalScope.launch {
            val result =
                    attendanceRecordDao.findDailyAttendanceByClazzUidAndDateAsync(currentClazzUid,
                            fromDate, toDate)

            //TODOne: KMP Manipulate this on UMCalendar side and fix this.
            for (everyDayAttendance in result!!) {
                //null time so that we're only working wiht the same day's group by
                val dd = everyDayAttendance.logDate
                val d = UMCalendarUtil.zeroOutTimeForGivenLongDate(dd)
                val a = everyDayAttendance.attendancePercentage
                lineDataMap[d.toFloat() / 1000] = a

            }

            //Remove messy date keys
            val ldpi = lineDataMap.entries.iterator()
            val lineDataMapFixedX = LinkedHashMap<Float, Float>()
            var l = 0f
            while (ldpi.hasNext()) {
                l++
                lineDataMapFixedX[l] = ldpi.next().value
            }

            view.updateAttendanceLineChart(lineDataMapFixedX)

        }

        //Calculate attendance average numbers for the bar chart.
        val clazzMemberDao = repository.clazzMemberDao

        GlobalScope.launch {
            val result =
                    clazzMemberDao.findAttendanceSpreadByThresholdForTimePeriodAndClazzAndType(
                            STATUS_ATTENDED, currentClazzUid, fromDate, toDate)
            barDataMap[3f] = result!!.high / 100
            barDataMap[2f] = result.mid / 100
            barDataMap[1f] = result.low / 100
            view.updateAttendanceBarChart(barDataMap)
        }

    }
}
